device/esp_tinyusb/descriptors_control.c

```diff
@@ -89,11 +160,94 @@ uint16_t const *tud_descriptor_string_cb(uint8_t index, uint16_t langid)
+    if (NULL == config->device_descriptor) {
+        ESP_LOGW(TAG, "The device's device descriptor is not provided by user, using default.");
+    }
+    s_desc_cfg.dev = config->device_descriptor
+                     ? config->device_descriptor
+                     : &descriptor_dev_kconfig;
+
+    // Select Configuration Descriptor
+#if (TUD_OPT_HIGH_SPEED)
+    // High Speed
+    if (NULL == config->hs_cfg_desc) {
+        // Default configuration descriptor is provided only for CDC, MSC and NCM classes
+#if (CFG_TUD_HID > 0 || CFG_TUD_MIDI > 0 || CFG_TUD_CUSTOM_CLASS > 0 || CFG_TUD_ECM_RNDIS > 0 || CFG_TUD_DFU > 0 || CFG_TUD_DFU_RUNTIME > 0 || CFG_TUD_BTH > 0)
+        ESP_GOTO_ON_FALSE(config->hs_cfg_desc, ESP_ERR_INVALID_ARG, fail, TAG, "Configuration descriptor must be provided for this device");
+#else
+        ESP_LOGW(TAG, "The device's HS configuration descriptor is not provided by user, using default.");
+#endif
+    }
+    s_desc_cfg.cfg = config->hs_cfg_desc
+                     ? config->hs_cfg_desc
+                     : descriptor_cfg_kconfig;
+
+    // Full Speed
+    if (NULL == config->fs_cfg_desc) {
+        // Default configuration descriptor is provided only for CDC, MSC and NCM classes
+#if (CFG_TUD_HID > 0 || CFG_TUD_MIDI > 0 || CFG_TUD_CUSTOM_CLASS > 0 || CFG_TUD_ECM_RNDIS > 0 || CFG_TUD_DFU > 0 || CFG_TUD_DFU_RUNTIME > 0 || CFG_TUD_BTH > 0)
+        ESP_GOTO_ON_FALSE(config->fs_cfg_desc, ESP_ERR_INVALID_ARG, fail, TAG, "Configuration descriptor must be provided for this device");
+#else
+        ESP_LOGW(TAG, "The device's FS configuration descriptor is not provided by user, using default.");
+#endif
+    }
+    s_desc_cfg.fs_cfg = config->fs_cfg_desc
+                        ? config->fs_cfg_desc
+                        : descriptor_fs_cfg_kconfig;
+
+    // HS and FS cfg desc should be equal length
+    ESP_GOTO_ON_FALSE(((tusb_desc_configuration_t *)s_desc_cfg.hs_cfg)->wTotalLength ==
+                      ((tusb_desc_configuration_t *)s_desc_cfg.fs_cfg)->wTotalLength,
+                      ESP_ERR_INVALID_ARG, fail, TAG, "HS and FS Configuration descriptors must be same length");
+
+    // Qualifier Descriptor
+    if (NULL == config->qualifier_descriptor) {
+        ESP_GOTO_ON_FALSE((s_desc_cfg.dev == &descriptor_dev_kconfig), ESP_ERR_INVALID_ARG, fail, TAG, "Qualifier Descriptor must be present (Device Descriptor not default).");
+        // Get default qualifier if device descriptor is default
+        ESP_LOGW(TAG, "The device's qulifier descriptor is not provided by user, using default.");
+        s_desc_cfg.qualifier = &descriptor_qualifier_kconfig;
+    } else {
+        s_desc_cfg.qualifier = config->qualifier_descriptor;
+    }
+
+    // Other Speed buffer allocate
+    s_desc_cfg.other_speed = calloc(1, ((tusb_desc_configuration_t *)s_desc_cfg.hs_cfg)->wTotalLength);
+    ESP_GOTO_ON_FALSE(s_desc_cfg.other_speed, ESP_ERR_NO_MEM, fail, TAG, "Other speed memory allocation error");
+#else
```

Comment:
Nitpick: For large `#if`/`#else`/`#endif` blocks, suggesting adding a comment of the `#if` condition to make the code more readable.

```suggestion
#else // TUD_OPT_HIGH_SPEED
```